ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -590,140 +590,270 @@ private boolean copyDataByIndexSegment(LogSegment logSegmentToCopy, IndexSegment
+              addAllEntriesForKeyInSegment(validEntries, indexSegment, indexEntry);
+            } else {
+              logger.trace("{} in index segment with start offset {} in {} is not valid because it is a deleted PUT",
+                  indexEntry, indexSegment.getStartOffset(), storeId);
+            }
           } else {
-            logger.trace("Skipping {} in index segment with start offset {} in {} because it is a deleted PUT",
-                indexEntry, indexSegmentStartOffset, storeId);
+            // valid PUT entry
+            // Add all values in this index segment (to account for the presence of TTL updates)
+            addAllEntriesForKeyInSegment(validEntries, indexSegment, indexEntry);
           }
         } else {
-          // valid PUT entry
-          validEntries.add(indexEntry);
+          logger.trace("{} in index segment with start offset {} in {} is not valid because it is an expired PUT",
+              indexEntry, indexSegment.getStartOffset(), storeId);
         }
-      } else {
-        logger.trace("Skipping {} in index segment with start offset {} in {} because it is an expired PUT", indexEntry,
-            indexSegmentStartOffset, storeId);
       }
     }
     return validEntries;
   }
 
   /**
-   * Checks if a record already exists in the target log (swap spaces).
+   * Determines whether a TTL update entry is valid
+   * @param key the {@link StoreKey} being examined
+   * @param indexSegmentStartOffset the start offset of the {@link IndexSegment} that the TTL update record is in
+   * @return {@code true} if the TTL update entry is valid
+   * @throws StoreException if there are problems reading the index
+   */
+  private boolean isTtlUpdateEntryValid(StoreKey key, Offset indexSegmentStartOffset) throws StoreException {
+    boolean valid = false;
+    //  A TTL update entry is "valid" if the corresponding PUT is still alive
+    // The PUT entry, if it exists, must be "before" this TTL update entry.
+    FileSpan srcSearchSpan = new FileSpan(srcIndex.getStartOffset(), indexSegmentStartOffset);
+    IndexValue srcValue = srcIndex.findKey(key, srcSearchSpan, EnumSet.of(PersistentIndex.IndexEntryType.PUT));
+    if (srcValue == null) {
+      // PUT is not in the source - therefore can't be in target. This TTL update can be cleaned up
+      logger.trace("TTL update of {} in segment with start offset {} in {} is not valid the corresponding PUT entry "
+          + "does not exist anymore", key, indexSegmentStartOffset, storeId);
+    } else {
+      IndexValue tgtValue = tgtIndex.findKey(key, null, EnumSet.of(PersistentIndex.IndexEntryType.PUT));
+      if (tgtValue == null && isOffsetInCurrentCycle(srcValue.getOffset())) {
+        // exists in src but not in tgt. This can happen either because
+        // 1. The FileSpan to which srcValue belongs is not under compaction
+        // 2. srcValue will be compacted in this cycle
```

Comment:
Hard to understand 2, more comments for 2 would be great.

Reply:
I added some more comments. Let me know if more info can be provided